feat: support configurable way to  retrieve recipes via github

lifeizhou-ap opened the pull request:
**Why**
Retrieve recipe via name from local directories and github, so that user can use shared recipe in github

**What**
Before this change, the user can only pass the recipe file path.  With this change, the logic will be

- If the user passes the recipe file path (eg: ~/Downloads/recipe.yaml),  it will render the file from the path
- If the user passes the recipe name (my-recipe), it will
      - search the my-recipe.yaml or my-recipe.json in the current directory
      - if "GOOSE_RECIPE_GITHUB_REPO" is set up in config.yaml (eg: `GOOSE_RECIPE_GITHUB_REPO=squareup/goose-recipes), search the recipe.yaml o recipe.json in the github.com/squareup/goose-recipes/my-recipe directory via git clie
  
This logic applies to `goose run --recipe`, `goose recipe deeplink` `goose recipe validate` cli


- Added the options to set GOOSE_RECIPE_GITHUB_REPO to Goose settings in `goose configuration` cli

![Uploading Screenshot 2025-05-02 at 4.33.17 pm.png…]()

DOsinga reviewed (commented):
This is great. I think we should just start with shelling out to git though, just do a git clone, in a hidden folder and pull from there from time to time

lifeizhou-ap commented on crates/goose-cli/src/recipes/github_recipe.rs:
```diff
@@ -0,0 +1,78 @@
+
+        // GitHub API URL for the recipe file
+        let api_url = format!(
+            "https://api.github.com/repos/squareup/goose-recipes/contents/{}/recipe.{}?ref=douwe/joke-of-the-day",
```
yes. it was just for testing.

kalvinnchau reviewed (changes requested):
We don't want any internal specific logic built in the the OSS app, that includes the git url and any other environment variables.

Automatically searching the user's various directories and then searching remote endpoints is unexpected behavior for a flag like that and shouldn't be automatically done. 

Imo when a user passes in a `--recipe` the expectation is to use the file and not go searching around.
If we want to support automatically searching, I think it should be a separate, explicit `--search` flag, even then the expectations are unclear around what/where it looks. 
It should be well defined to the user that it will go looking around their computer; and imo it shouldn't automatically go searching for remote files (or split it into `--search-remote $URL`).

If we want remote support it should be generic - i.e. `--url`  that supports various types of endpoints `https://` or `git://` not specific to any org.

DOsinga commented:
So me and @kalvinnchau had a quick chat about this, here's our rough consensus:

* We can't make changes that impact the open source version in a weird way like adding a fixed github repository
* We do want to make recipes discoverable

So let's do this:
* Remove the ~/goose_recipes search for now. We might bring this back if useful
* Let's add an optional setting in goose settings that specifies the github repository to search in
* Let's add this to the install script and then later to sq goose (currently pointing to the python version, ugh)


i.e. we now have:

    curl -fsSL https://github.com/block/goose/releases/download/stable/download_cli.sh \
    | GOOSE_PROVIDER=databricks GOOSE_MODEL=goose \
    DATABRICKS_HOST=https://block-lakehouse-production.cloud.databricks.com/ bash

change that to:

    curl -fsSL https://github.com/block/goose/releases/download/stable/download_cli.sh \
    | GOOSE_PROVIDER=databricks GOOSE_MODEL=goose \
    RECIPE_GIT=square_up/goose-recipes \
    DATABRICKS_HOST=https://block-lakehouse-production.cloud.databricks.com/ bash

lifeizhou-ap commented:
> So me and @kalvinnchau had a quick chat about this, here's our rough consensus:
> 
> * We can't make changes that impact the open source version in a weird way like adding a fixed github repository
> * We do want to make recipes discoverable
> 
> So let's do this:
> 
> * Remove the ~/goose_recipes search for now. We might bring this back if useful
> * Let's add an optional setting in goose settings that specifies the github repository to search in
> * Let's add this to the install script and then later to sq goose (currently pointing to the python version, ugh)
> 
> i.e. we now have:
> 
> ```
> curl -fsSL https://github.com/block/goose/releases/download/stable/download_cli.sh \
> | GOOSE_PROVIDER=databricks GOOSE_MODEL=goose \
> DATABRICKS_HOST=https://block-lakehouse-production.cloud.databricks.com/ bash
> ```
> 
> change that to:
> 
> ```
> curl -fsSL https://github.com/block/goose/releases/download/stable/download_cli.sh \
> | GOOSE_PROVIDER=databricks GOOSE_MODEL=goose \
> RECIPE_GIT=square_up/goose-recipes \
> DATABRICKS_HOST=https://block-lakehouse-production.cloud.databricks.com/ bash
> ```

Hey team,

I have updated this PR.  

- Added the input for user to enter the recipe github name and it will saved as GOOSE_RECIPE_GITHUB_REPO in config.yaml file.  If the user passes the GOOSE_RECIPE_GITHUB_REPO environment variable, the env variable value will be displayed the default value in the input prompt.  If this environment variable is not passed, it will show the current value in the config.yaml file as the default value if it has been configured.

- I changed to use gh clone instead of git clone, because using gh I can pass the repo full name (org_name/repo_name) instead of the ssh clone url (org-49461806@github.com:squareup/goose-recipes.git) which is a bit customised.

kalvinnchau reviewed (changes requested):
* we'll need to document somewhere that `gh` is a required dependency for using the github recipe feature
  * maybe as a note in the configure step? 
* can we make `GOOSE_RECIPE_GITHUB_REPO` a `const` in `crates/goose-cli/src/recipes/github_recipe.rs` and re-used in the various places its hard-coded at the moment
* there's a couple uses of `.expect(..)` can we use a `map_err` approach and print the resulting error rather than having it panic

lifeizhou-ap commented on crates/goose-cli/src/recipes/github_recipe.rs:
```diff
@@ -0,0 +1,131 @@
+}
+
+fn ensure_repo_cloned(recipe_repo_full_name: &str) -> Result<PathBuf> {
+    let local_repo_parent_path = env::temp_dir();
```
this is temp directory based on the platform https://doc.rust-lang.org/std/env/fn.temp_dir.html for example: macos: /tmp.  It should exist already. It won't create unless it does not exist.  

User has to clone the repo at least once before git fetch. We use temp directory as the parent dir when the user clones the repo.

kalvinnchau replied:
Ah yes, sorry I confused it with the `tempfile::tempdir` which we use in other places which deletes the directory after the tempdir binding is out of scope.

lifeizhou-ap commented on crates/goose-cli/src/recipes/github_recipe.rs:
agree. I've added the message

.map_err(|_| anyhow::anyhow!("Failed to run `gh auth status`. Make sure you have `gh` installed."))?;

lifeizhou-ap commented on crates/goose-cli/src/commands/configure.rs:
```diff
@@ -1104,3 +1112,33 @@ pub async fn configure_tool_permissions_dialog() -> Result<(), Box<dyn Error>> {
+        }
+    }
+    let mut recipe_repo_input = cliclack::input(
+        "Enter the Goose Recipe Github Repo in the format of owner/repo: eg squareup/goose-recipes",
```
It might confuse people if we use `block/goose-recipes` because this repo does not exist.  Currently `goose-recipes` is only in `squareup` and probably some of the recipes are not good to expose to a public repo. 

I guess in the future we might have `block/goose-recipes` in the open source community.  For now. I may change 

`Enter the Goose Recipe Github Repo in the format of owner/repo: eg squareup/goose-recipes` =>  `Enter your Goose Recipe Github repo (owner/repo): e.g. squareup/goose-recipes`

How about that?

lifeizhou-ap commented on crates/goose-cli/src/commands/configure.rs:
```diff
@@ -1104,3 +1112,33 @@ pub async fn configure_tool_permissions_dialog() -> Result<(), Box<dyn Error>> {
+        recipe_repo_input = recipe_repo_input.default_input(&recipe_repo);
+    }
+    let input_value: String = recipe_repo_input.interact()?;
+    if input_value.clone().trim().is_empty() {
```
added the comment
